x-pack/plugins/timelines/server/search_strategy/timeline/factory/helpers/format_timeline_data.ts

```diff
@@ -5,118 +5,116 @@
-  dataFields: readonly string[],
-  ecsFields: readonly string[],
-  hit: EventHit
-) =>
-  uniq([...ecsFields, ...dataFields]).reduce<Promise<TimelineEdges>>(
-    async (acc, fieldName) => {
-      const flattenedFields: TimelineEdges = await acc;
-      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-      flattenedFields.node._id = hit._id!;
-      flattenedFields.node._index = hit._index;
-      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-      flattenedFields.node.ecs._id = hit._id!;
-      flattenedFields.node.ecs.timestamp = getTimestamp(hit);
-      flattenedFields.node.ecs._index = hit._index;
-      if (hit.sort && hit.sort.length > 1) {
-        flattenedFields.cursor.value = hit.sort[0];
-        flattenedFields.cursor.tiebreaker = hit.sort[1];
-      }
-      const waitForIt = await mergeTimelineFieldsWithHit(
-        fieldName,
-        flattenedFields,
-        hit,
-        dataFields,
-        ecsFields
-      );
-      return Promise.resolve(waitForIt);
-    },
-    Promise.resolve({
-      node: { ecs: { _id: '' }, data: [], _id: '', _index: '' },
-      cursor: {
-        value: '',
-        tiebreaker: null,
-      },
-    })
-  );
-
-const getValuesFromFields = async (
+const createBaseTimelineEdges = (): TimelineEdges => ({
+  node: {
+    ecs: { _id: '' },
+    data: [],
+    _id: '',
+    _index: '',
+  },
+  cursor: {
+    value: '',
+    tiebreaker: null,
+  },
+});
+
+const processMetadataField = (fieldName: string, hit: EventHit): TimelineNonEcsData[] => [
+  {
+    field: fieldName,
+    value: toStringArray(get(fieldName, hit)),
+  },
+];
+
+const processFieldData = (
   fieldName: string,
   hit: EventHit,
   nestedParentFieldName?: string
-): Promise<TimelineNonEcsData[]> => {
-  if (ECS_METADATA_FIELDS.includes(fieldName)) {
-    return [{ field: fieldName, value: toStringArray(get(fieldName, hit)) }];
-  }
+): TimelineNonEcsData[] => {
+  const fieldToEval = nestedParentFieldName
+    ? { [nestedParentFieldName]: hit.fields[nestedParentFieldName] }
+    : { [fieldName]: hit.fields[fieldName] };
 
-  let fieldToEval;
+  const formattedData = getDataFromFieldsHits(fieldToEval);
 
-  if (nestedParentFieldName == null) {
-    fieldToEval = {
-      [fieldName]: hit.fields[fieldName],
-    };
-  } else {
-    fieldToEval = {
-      [nestedParentFieldName]: hit.fields[nestedParentFieldName],
-    };
-  }
-  const formattedData = await getDataSafety(getDataFromFieldsHits, fieldToEval);
-  return formattedData.reduce((acc: TimelineNonEcsData[], { field, values }) => {
-    // nested fields return all field values, pick only the one we asked for
-    if (field.includes(fieldName)) {
-      acc.push({ field, value: values });
-    }
-    return acc;
-  }, []);
+  // Filter and map in one pass
+  return formattedData
+    .filter((item) => item.field.includes(fieldName))
+    .map(({ field, values }) => ({
+      field,
+      value: values,
+    }));
```

Comment:
```suggestion
  return formattedData.reduce((agg, { field, values }) => {
    if (field.includes(fieldName)) {
      agg.push({
        field,
        value: values,
      });
    }
    return agg;
  }, []);

```

Using reduce here, you can save one iteration of the array